src/modules/Elsa.Workflows.Api/Endpoints/WorkflowInstances/Cancel/Endpoint.cs

```diff
@@ -0,0 +1,45 @@
+using Elsa.Abstractions;
+using Elsa.Workflows.Runtime.Contracts;
+using Elsa.Workflows.Runtime.Models;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Http;
+
+namespace Elsa.Workflows.Api.Endpoints.WorkflowInstances.Cancel;
+
+[PublicAPI]
+internal class Cancel : ElsaEndpoint<Request>
+{
+    private readonly IWorkflowRuntime _workflowRuntime;
+
+    public Cancel(IWorkflowRuntime workflowRuntime)
+    {
+        _workflowRuntime = workflowRuntime;
+    }
+
+    public override void Configure()
+    {
+        Post("/cancel/workflow-instances/{id}");
+        ConfigurePermissions("cancel:workflow-instances");
+        //Allows for post with empty body
+        Description(x => x.Accepts<Request>("*/*"), clearDefaults: true);
+    }
+    
+    public override async Task HandleAsync(Request request, CancellationToken cancellationToken)
+    {
+        var result = await _workflowRuntime.CancelWorkflowAsync(request.Id, cancellationToken);
+        
+        if (result.Result)
+            await SendOkAsync(cancellationToken);
+        else if (result.Reason == FailureReason.NotFound)
+            await SendNotFoundAsync(cancellationToken);
+        else
+        {
+            if (result.Reason == FailureReason.InvalidState)
+                AddError("Instance is in finished state.");
+            else
+                AddError("Unable to access instance.");
+            
+            await SendErrorsAsync(StatusCodes.Status409Conflict, cancellationToken);
```

Comment:
I am having second thoughts if we should return a bad request response. Let's consider the following scenario:

The user decides that they want to abort a bulk operation and cancel everything. Some of the dispatched workflows may have already completes. The user doesn't want their API call to fail because of that.

Perhaps it makes sense to just respond with a "report" (as a simple JSON model) of what got cancelled and what was already cancelled.

What do you think?

Reply:
The bad request is only returned on this endpoint (for a single instance cancellation. If the given instance can be found it will give a succeeded result regardless what happens with the child workflows.
In the case of multiple instance cancellation only the amount of workflows that should be cancelled is reported (no bad requests). The only way to get a non 200 on that endpoint is if the validation fails or if there is an uncaught exception.

Reply:
Gotcha. then I agree 👍🏻